armsrc/Standalone/dankarmulti.c

```diff
@@ -0,0 +1,138 @@
+//-----------------------------------------------------------------------------
+// Daniel Karling, 2021
+//
+// This code is licensed to you under the terms of the GNU GPL, version 2 or,
+// at your option, any later version. See the LICENSE.txt file for the text of
+// the license.
+//-----------------------------------------------------------------------------
+// main code for Multi Loader
+//-----------------------------------------------------------------------------
+#include "standalone.h" // standalone definitions
+#include "proxmark3_arm.h"
+#include "appmain.h"
+#include "fpgaloader.h"
+#include "ticks.h"
+#include "util.h"
+#include "dbprint.h"
+
+/*
+ * This mode will allow you to access multiple different standalone
+ * modes at the same time. The below section defines which modes are
+ * available. Modes are added to the mode list according to the
+ * following:
+ *
+ *  1) Include "dankarmulti.h"
+ *  2) Define MODE_NAME and MODE_FILE for a mode (MODE_NAME can
+ *     of your choosing, but must be a valid C token. I.e. pretend
+ *     that you are naming a variable)
+ *  3) Include "dankarmulti.h" again
+ *  4) Repeat steps 2 and 3 for additional modes.
+ *  5) Use the macros START_MODE_LIST, ADD_MODE and END_MODE_LIST
+ *     to create the list of modes. You need to use the same names
+ *     here as defined earlier.
+ *
+ *  Usage:
+ *  Single press to cycle between the modes.
+ *  The LEDs will show the currently selected mode.
+ *  Hold button to start the selected mode.
+ *
+ *  How many modes can you have at the same time? Depends on memory,
+ *  but the LEDs will overflow after 15.
+ *
+ *  I don't know if this works with all the different standalone modes.
+ *  I would imagine that it doesn't. If two modes are included with
+ *  functions that collide (name-wise) there will be issues.
+ *
+ *  NOTE: You will have to keep track of if the included modes require
+ *  external memory or bluetooth yourself. The mode selection in
+ *  the Makefiles is not able to do it.
+ */
+
+
+/*******************
+ * Begin mode list *
+ *******************/
+
+#include "dankarmulti.h"
+
+#define MODE_NAME sniff14a
+#define MODE_FILE "hf_14asniff.c"
+#include "dankarmulti.h"
+#define MODE_NAME em4100
+#define MODE_FILE "lf_em4100rswb.c"
+#include "dankarmulti.h"
+#define MODE_NAME icehid
+#define MODE_FILE "lf_icehid.c"
+#include "dankarmulti.h"
```

Comment:
question,
how does re-defining the same macro work in the add_mode down below?

Reply:
The same name is used again for ADD_MODE simply to add the different modes to a list. It simplifies enumerating the included modes.

Reply:
yes,  but the redefine is done before the ADD_MODE macro..

Reply:
The macros are undef:d in the "dankarmulti.h" header, at the end of it, meaning they can be redefined for the next mode.

Reply:
.... sneaky way of dealing with macro defines...   Not sure about it,